

library(vegan)
library(ggplot2)
library(ggfortify)

############################################################
# Various colour colour_palettes
my_colour_palette <- c("#8dd3c7","#ffffb3","#bebada","#fb8072", "#80b1d3", "#fdb462","#b3de69","#fccde5","#d9d9d9","#bc80bd","#ccebc5", "#cc0000")
# From http://tools.medialab.sciences-po.fr/iwanthue/
my_colour_palette_20 <- c("#66bd79","#a35bcf","#5bb643","#d14ea6","#a2b239","#5c6bcc","#dc892e","#5e93cd","#d64737","#49b6a8","#dc3c6e","#4f7e3c","#bd8cd5","#caab55","#914c88","#867230","#df82a2","#a65429","#ab4a5a","#e0896a")
my_colour_palette_20_distinct <- c("#0057b4","#7fff56","#d600bc","#d8d500","#e76eff","#019932","#9f8fff","#ffc730","#007fac","#a20019","#06fefd","#ff6782","#00774c","#e0c8ff","#717a00","#4b2952","#e2ed7d","#46321e","#ffbd76","#ffb4c6")
my_colour_palette_30_distinct <- c("#009348","#f579fe","#4fe16e","#b40085","#4d7e00","#4742b4","#f0c031","#016dd9","#d45200","#7499ff","#ef4d2d","#01c9c8","#f8394b","#88d7a6","#d20063","#c8cc5d","#882986","#fdb95d","#404f8f","#917300","#f3aefc","#5c5800","#ff75c3","#00674a","#ba001c","#979760","#8b354c","#ff875f","#943105","#cf9478")
my_colour_palette_206_distinct <- c("#cfefb4","#7d8b00","#a70079","#552155","#632900","#ffb173","#fbdcf2","#015a6a","#43fdf7","#ff443a","#008186","#3b8aff","#8b5fff","#ff9777","#4200a9","#85f6fd","#c96000","#36218a","#d28900","#0137d7","#30325b","#ff836b","#008b4f","#21ff9d","#00794d","#870052","#e9ec4b","#ce006b","#6e0044","#8a6500","#006971","#432e4b","#ca8dff","#f20059","#44ffe2","#00be5c","#a0d2ff","#1914ab","#4d284e","#59d7ff","#ab9aff","#0151d9","#1de740","#e24500","#9fc400","#610769","#0a4600","#1e365b","#018f3f","#b15fff","#009c5e","#005290","#506100","#f49aff","#0187c1","#ffb5f4","#daf100","#70081d","#ff9890","#c1baff","#ffbe5a","#1b3466","#ff2a7f","#ff5d3c","#e47800","#ac6bff","#1f6000","#006627","#4f4000","#dcd6ff","#ffd7c1","#ed2de4","#a50038","#a5a8ff","#0f2f7f","#b11700","#00e06b","#ffabb8","#015780","#82eaff","#1b2a88","#6f1600","#d3ef9c","#746e00","#01d851","#625300","#01d799","#96fd6c","#ff5ca1","#7b0017","#004c2b","#baf678","#f8aaff","#007c1b","#01a88a","#a71ed8","#fb8cff","#840079","#276d00","#556655","#02b0de","#c0efd7","#63193e","#8e9984","#017ac9","#ff925f","#ff63d7","#294100","#28baff","#5b2523","#35ab00","#69132e","#8a3b00","#a67700","#7fff6a","#002f96","#681a0b","#4d3003","#ff7de6","#0190d8","#a69700","#ff6282","#d3f266","#ffc4cf","#ffac3c","#d064ff","#d07aff","#c3005d","#9d0067","#0167c1","#8cfe82","#ffd68f","#8cfcaf","#f50096","#00c2a2","#aa5e00","#02c16d","#4e4bf6","#ffd962","#004793","#93d800","#462a58","#323a03","#4f9eff","#2b3a25","#2defff","#02edd6","#864e00","#ffc59f","#e7e9ab","#014cc4","#437bff","#00afba","#ff7d82","#8a1ed4","#ff48b3","#acf7ab","#005550","#7600a6","#bc0028","#00adab","#02dfbf","#ba004c","#004760","#ebc5ff","#0162d7","#9b3900","#5869ff","#ff6160","#87b6ff","#ff6796","#ff8422","#ff8440","#b500a8","#937fff","#0132bd","#f48e00","#1e8800","#462370","#3e3614","#9ca800","#efe5bf","#aeb6a0","#d9aaff","#d8ef89","#cec800","#ffb8b3","#4a2c42","#01715b","#b8ebff","#ff9ec0","#ff93ec","#ffe0aa","#65b300","#6a8b00","#f6e77c","#ff85c0","#5de522","#a5f6ca","#c70077","#5a4149","#a3b700","#ff63c4","#63fecd","#93f6e7","#01b4a4")
my_colour_palette_15 <- c("#77b642","#7166d9","#cfa240","#b351bb","#4fac7f","#d44891","#79843a","#c68ad4","#d15a2c","#5ba7d9","#ce4355","#6570ba","#b67249","#9b4a6f","#df8398")
my_colour_palette_32_distinct <- c("#ea7e00","#ca0074","#d1c69b","#474007","#bb00ad","#9c80ff","#be3300","#542e72","#00b9f5","#09436b","#8b0036","#9ac8e6","#ff1059","#959eff","#154a11","#0290f4","#ff7762","#7dbf00","#ff8194","#834c00","#006e73","#f9bb5d","#d6c943","#017229","#00d3a8","#732427","#36e191","#6a8200","#efb3ea","#3227bb","#ff90e1","#e92a12")
lesion_colour_palette_7 <- c("#8558d6","#6ee268","#d247ad","#c9d743","#d7453e","#59a237","#d78f2a")
patient_colour_palette_45 <- c("#d64530","#585fb1","#795d97","#9e4773","#3f6921","#71692c","#a2b93c","#d571cc","#9b3e97","#33947a","#98ad66","#448a4e","#869ae0","#5ce7af","#e085a3","#dfdc87","#d19be2","#5cb735","#e38269","#3db6c0","#50b565","#50902c","#a98a2c","#dde84a","#db3d76","#5fe485","#7c8329","#b3e791","#6fe965","#5ebce9","#3c86c1","#2a6a45","#65b688","#6651d1","#af4ed3","#df872f","#56e4db","#737cea","#ac464b","#dd37b5","#995b2b","#daac6f","#92e2be","#a2e24b","#e0be3a")
my_colour_palette_10_distinct <- c("#8eec45","#0265e8","#f6a800","#bf6549","#486900","#c655a0","#00d1b6","#ff4431","#aeb85c","#7e7fc8")
######################## Functions #########################

# Function that calculates the geometric mean with some error-protection bits. 
# DESeq2 does not appear to work (will throw an error) if every OTU (or genus or genome etc.) 
# contains at least one count of zero in every row of the count data.
# Specifically, the function "dds<-DESeq(dds, betaPrior = FALSE)" will fail
# One way to address this is to use the function below as input to DESeq2 to transform the data.
# Calculate the geometric means prior to estimating the size factors
gm_mean = function(x, na.rm=TRUE){
  # The geometric mean, with some error-protection bits.
  exp(sum(log(x[x > 0 & !is.na(x)]), na.rm=na.rm) / length(x))
}

# Center log ratio transform
clr = function(x, base=2){
  x <- log((x / gm_mean(x)), base)
  x[!is.finite(x) | is.na(x)] <- 0.0
  return(x)
}

# For each rowname (OTU), get the corresponding taxonomy_species
# Assumes "OTU.ID" and "taxonomy_species" columns in the provided map dataframe
assign_taxonomy_to_otu <- function(otutable, taxon_map){
  taxonomies <- c()
  for (otuid in rownames(otutable)){
    taxonomies <- c(taxonomies, as.character(taxon_map[taxon_map$OTU.ID == otuid,]$taxonomy_species))
  }
  return(taxonomies)
}

# Function that takes the metadata and a list of variables (column names) and returns those samples (rownames) with NA entries
get_samples_missing_data <- function(my_metadata, variables){
  samples_missing_data <- c()
  for (name in variables) {
    samples_missing_data <- c(samples_missing_data, rownames(my_metadata[is.na(my_metadata[[name]]),]))
  }
  return(unique(samples_missing_data))
}
############################################################
# Set the working directory
setwd("/Users/julianzaugg/Desktop/ACE/major_projects/skin_microbiome_16S")

# Load count table at the OTU level. These are the counts for OTUs that were above our abundance thresholds
otu_rare.df <- read.table("Result_tables/count_tables/OTU_counts_rarefied.csv", sep =",", header =T)
otu.df <- read.table("Result_tables/count_tables/OTU_counts.csv", sep =",", header =T)

# Load the OTU - taxonomy mapping file
otu_taxonomy_map.df <- read.csv("Result_tables/other/otu_taxonomy_map.csv", header = T)

# Load the processed metadata
metadata.df <- read.csv("Result_tables/other/processed_metadata.csv", sep =",", header = T)

# metadata.df <- metadata.df[metadata.df$Project == "immunocompetent",]
# metadata.df <- metadata.df[metadata.df$Project == "immunocompromised",]

# metadata.df <- metadata.df[!metadata.df$Patient %in% c("MS001","MS010"),]

# Set the Index to be the rowname
rownames(metadata.df) <- metadata.df$Index

# Since we likely removed samples from the count matrix
# in the main script, remove them from the metadata.df here
samples_removed <- metadata.df$Index[!metadata.df$Index %in% names(otu_rare.df)]
metadata.df <- metadata.df[! metadata.df$Index %in% samples_removed,]

# Factorise discrete columns
metadata.df$Patient <- factor(metadata.df$Patient)
metadata.df$Sampletype <- factor(metadata.df$Sampletype)
metadata.df$Project <- factor(metadata.df$Project)
metadata.df$Patient_group <- factor(metadata.df$Patient_group)
metadata.df$Gender <- factor(metadata.df$Gender)
metadata.df$Number_of_meds <- factor(metadata.df$Number_of_meds)

# We are only interested in C,AK_PL,IEC_PL,SCC_PL,AK,IEC and SCC lesions. 
# Remove samples for different lesion types (nasal,scar,scar_PL,KA,KA_PL,VV,VV_PL,SF,SF_PL,other,other_PL) from metadata and otu table
metadata.df <- metadata.df[metadata.df$Sampletype %in% c("C","AK_PL","IEC_PL","SCC_PL","AK","IEC","SCC", "NLC"),]

# Remove samples from the OTU table that are not in the filtered metadata
otu_rare.df <- otu_rare.df[,names(otu_rare.df) %in% c("OTU.ID", as.character(metadata.df$Index))]
otu.df <- otu.df[,names(otu.df) %in% c("OTU.ID", as.character(metadata.df$Index))]

# Order the metadata.df by the index value
# metadata.df <- metadata.df[order(metadata.df$Index),]

# Create matrices
otu_rare.m <- otu_rare.df
rownames(otu_rare.m) <- otu_rare.df$OTU.ID
otu_rare.m$OTU.ID <- NULL
otu_rare.m <- as.matrix(otu_rare.m)

otu.m <- otu.df
rownames(otu.m) <- otu.df$OTU.ID
otu.m$OTU.ID <- NULL
otu.m <- as.matrix(otu.m)

# Filter by reads per sample if you don't want to use the existing filtering
minimum_reads <- 0
otu_rare.m <- otu_rare.m[,colSums(otu_rare.m) >= minimum_reads]
otu.m <- otu.m[,colSums(otu.m) >= minimum_reads]
metadata.df <- metadata.df[rownames(metadata.df) %in% colnames(otu_rare.m),]

# Order the matrices and metadata to be the same order
metadata.df <- metadata.df[order(rownames(metadata.df)),]
otu_rare.m <- otu_rare.m[,order(rownames(metadata.df))]
otu.m <- otu.m[,order(rownames(metadata.df))]
# colnames(otu_rare.m) == rownames(metadata.df)

# Just get OTUs with more than counts of 15 (0.05 % if rarefied to 30,000)
# Just get OTUs with more than counts of 3 (0.01 % if rarefied to 30,000)
dim(otu_rare.m)
otu_rare_filtered.m <- otu_rare.m[apply(otu_rare.m,1,max) >= 15,]
dim(otu_rare_filtered.m)

dim(otu.m)
otu_filtered.m <- otu_rare.m[apply(otu.m,1,max) >= 15,]
dim(otu_filtered.m)

# CLR transform the otu matrix.
otu_rare_clr_filtered.m <- clr(otu_rare_filtered.m)
otu_clr_filtered.m <- clr(otu_filtered.m)

# If there are negative values, assign them a value of zero
otu_rare_clr_filtered.m[which(otu_rare_clr_filtered.m < 0)] <- 0
otu_clr_filtered.m[which(otu_clr_filtered.m < 0)] <- 0

# Determine which samples are missing metadata and remove them
# variables_of_interest <- c("Sampletype", "Patient","Sampletype_pooled", "Project", "Sampletype_pooled_IEC_sep", "Sampletype_pooled_C_sep")
# samples_to_remove <- get_samples_missing_data(metadata.df, variables_of_interest)
# metadata.df <- metadata.df[!rownames(metadata.df) %in% samples_to_remove,]
#metadata.df <- metadata.df["Sampletype"]
# metadata.df <- metadata.df[variables_of_interest]

# --------------------------------------------------------------------------------
# --------------------------------------------------------------------------------
# --------------------------------------------------------------------------------
# Ordination analysis
generate_pca <- function(pca_object, mymetadata, variable_to_plot, colour_palette, limits = NULL, filename = NULL, include_legend = T, add_spider = F, add_ellipse = F,
                         point_alpha = 1, plot_width = 10, plot_height=10, point_size = 0.8, point_line_thickness = 1,
                         label_sites = F, label_species = F,
                         legend_x = NULL, legend_y = NULL, legend_x_offset = 0, legend_y_offset = 0,
                         plot_spiders = NULL, plot_ellipses = NULL,plot_hulls = NULL, legend_cols = 2, legend_title = NULL,
                         label_ellipse = F, ellipse_label_size = 0.5, ellipse_border_width = 1,variable_colours_available = F, 
                         plot_title = NULL, use_shapes = F){
  pca.scores <- try(scores(pca_object, choices=c(1,2,3)))
  
  if(inherits(pca.scores, "try-error")) {
    return()
  }
  # Get component x,y coordinates
  pca_site_scores <- scores(pca_object, display = "sites")
  pca_specie_scores <- scores(pca_object, display = "species")
  pca_percentages <- (pca_object$CA$eig/sum(pca_object$CA$eig)) * 100
  
  # Remove NA entries from the metadata and from the PCA
  internal_metadata <- mymetadata[!is.na(mymetadata[[variable_to_plot]]),]
  pca_site_scores <- pca_site_scores[rownames(pca_site_scores) %in% rownames(internal_metadata),]
  
  if (!is.null(limits)){
    x_min <- limits[1]
    x_max <- limits[2]
    y_min <- limits[3]
    y_max <- limits[4]
  }
  else {
    x_min <- round(lapply(min(pca_site_scores[,1]), function(x) ifelse(x > 0, x + 1, x - 1))[[1]])
    x_max <- round(lapply(max(pca_site_scores[,1]), function(x) ifelse(x > 0, x + 1, x - 1))[[1]])
    y_min <- round(lapply(min(pca_site_scores[,2]), function(x) ifelse(x > 0, x + 1, x - 1))[[1]])
    y_max <- round(lapply(max(pca_site_scores[,2]), function(x) ifelse(x > 0, x + 1, x - 1))[[1]])
    
  }
  
  
  my_xlab = paste("PC1 (", round(pca_percentages[1],1), "%)", sep = "")
  my_ylab = paste("PC2 (", round(pca_percentages[2],1), "%)", sep = "")
  
  metadata_ordered.df <- internal_metadata[order(rownames(internal_metadata)),]
  metadata_ordered.df <- metadata_ordered.df[order(metadata_ordered.df[[variable_to_plot]]),]
  
  if (!is.null(filename)){
    pdf(filename, height=plot_height,width=plot_width)  
  }
  
  plot(pca_object,
       type='n',
       # x = 0, y=0,
       xlim = c(x_min,x_max),
       ylim = c(y_min,y_max),
       xlab = my_xlab,
       ylab = my_ylab)
  
  # Make grid
  grid(NULL,NULL, lty = 2, col = "grey80")
  
  # Assign (unique) colours and shapes for each grouping variable
  # variable_values <- factor(sort(as.character(unique(metadata_ordered.df[[variable_to_plot]]))))
  variable_values <- factor(as.character(unique(metadata_ordered.df[[variable_to_plot]])))
  
  # If variable colour column "variable_colour" in metadata, use colours from there
  if (variable_colours_available == T){
    colour_col_name <- paste0(variable_to_plot, "_colour")
    variable_colours <- setNames(as.character(unique(metadata_ordered.df[[colour_col_name]])), as.character(unique(metadata_ordered.df[[variable_to_plot]])))
  } else{
    variable_colours <- setNames(colour_palette[1:length(variable_values)], variable_values)  
  }
  if (use_shapes == T){
    # variable_shapes <- setNames(c(25,24,23,22,21,8,6,5,4,3,2,1)[1:length(variable_values)],variable_values)
    variable_shapes <- setNames(rep(c(25,24,23,22,21),length(variable_values))[1:length(variable_values)],variable_values)
  } else{
    variable_shapes <- setNames(rep(c(21),length(variable_values))[1:length(variable_values)],variable_values)  
  }
  #variable_shapes <- setNames(c(25,24,23,22,21,8,6,5,4,3,2,1)[1:length(variable_values)],variable_values)
  # print(variable_colours)
  annotation_dataframe <- data.frame(variable_colours, variable_shapes)
  annotation_dataframe$variable_outline_colours <- as.character(annotation_dataframe$variable_colours)
  annotation_dataframe[annotation_dataframe$variable_shapes > 15,"variable_outline_colours"] <- "black"
  
  # Order the site scores by the order of the rows in the metadata
  pca_site_scores <- pca_site_scores[rownames(metadata_ordered.df),]
  
  all_sample_colours <- as.character(
    lapply(
      as.character(metadata_ordered.df[rownames(pca_site_scores),variable_to_plot]), 
      function(x) variable_colours[x]
    )
  )
  
  all_sample_shapes <- as.numeric(
    lapply(
      as.character(sort(metadata_ordered.df[rownames(pca_site_scores),variable_to_plot])), 
      function(x) variable_shapes[x][[1]]
    )
  )
  
  # Set the outline colours for all samples based on the sample colours and refering to the annotation dataframe created above
  all_sample_outline_colours <- as.character(unlist(lapply(all_sample_colours, function(x) annotation_dataframe[annotation_dataframe$variable_colours == x, "variable_outline_colours"])))
  
  # Need to construct the legend outline colour vector.
  legend_point_outline_colours <- annotation_dataframe$variable_outline_colours
  
  # for (i in 1:length(all_sample_colours)){ 
  #   if (as.numeric(all_sample_shapes[i]) < 15){
  #     
  #     all_sample_outline_colours <- c(all_sample_outline_colours, all_sample_colours[i])
  #   } else{
  #     all_sample_outline_colours <- c(all_sample_outline_colours, "black")
  #   }
  # }
  # print(all_sample_outline_colours)
  # all_sample_outline_colours
  
  # all_sample_outline_colours <- as.character(
  #   lapply(
  #     all_sample_shapes, 
  #     function(x) variable_colours[x]
  #   )
  # )
  # for (i in )
  # 
  
  points(pca_site_scores, 
         cex = point_size,
         lwd = point_line_thickness,
         pch = all_sample_shapes,
         col = alpha(all_sample_outline_colours,point_alpha),
         # col = alpha("black",point_alpha),
         bg = alpha(all_sample_colours, point_alpha),
  )
  
  # Plot ellipses that are filled
  plot_ellipses_func <- function () {
    for (member in variable_values) {
      if (nrow(metadata_ordered.df[metadata_ordered.df[[variable_to_plot]] == member,]) > 2){ # if too few samples, skip plotting ellipse
        ordiellipse(pca_site_scores,
                    groups = metadata_ordered.df[[variable_to_plot]],
                    kind = "ehull",
                    lwd = ellipse_border_width,
                    # border = variable_colours[member][[1]],
                    border = alpha(variable_colours[member][[1]],point_alpha),
                    # col = variable_colours[member][[1]],
                    col = alpha(variable_colours[member][[1]],point_alpha),
                    show.groups = member,
                    alpha = .05,
                    draw = "polygon",
                    label = F,
                    cex = .5)
      }
    }
  }
  
  # Plot hulls that are filled
  plot_hulls_func <- function () {
    for (member in variable_values){
      if (nrow(metadata_ordered.df[metadata_ordered.df[[variable_to_plot]] == member,]) > 2){ # if too few samples, skip plotting ellipse}
        ordihull(pca_site_scores,
                 groups = metadata_ordered.df[[variable_to_plot]],
                 lwd = ellipse_border_width,
                 # border = variable_colours[member][[1]],
                 border = alpha(variable_colours[member][[1]],point_alpha),
                 # col = variable_colours[member][[1]],
                 col = alpha(variable_colours[member][[1]],point_alpha),
                 show.groups = member,
                 alpha = .05,
                 draw = "polygon",
                 label = F,
                 cex = .5)
      }
    }
  }
  if (hasArg(plot_hulls)){
    if (plot_hulls == T){
      plot_hulls_func()    
    }
  }
  
  plot_ellipses_labels_func <- function(label_ellipse = F){
    # Repeat to have labels clearly on top of all ellipses
    for (member in variable_values){
      if (nrow(metadata_ordered.df[metadata_ordered.df[[variable_to_plot]] == member,]) > 2){ # if too few samples, skip plotting ellipse
        ordiellipse(pca_site_scores,
                    groups = metadata_ordered.df[[variable_to_plot]],
                    kind = "ehull",
                    # border = variable_colours[member][[1]],
                    border = NA,
                    # col = variable_colours[member][[1]],
                    col = NA,
                    show.groups = member,
                    alpha = 0,
                    draw = "polygon",
                    label = label_ellipse,
                    cex = ellipse_label_size)
      }
    }
  }
  
  if (hasArg(plot_ellipses) | hasArg(label_ellipse)){
    if (plot_ellipses == T){
      plot_ellipses_func()    
      plot_ellipses_labels_func(label_ellipse = label_ellipse)
    } else if (label_ellipse == T){
      plot_ellipses_labels_func(label_ellipse = label_ellipse)
    }
  } 
  
  #Plot spiders
  plot_spiders_func <- function (label_spider = F) {
    for (member in variable_values){
      if (nrow(metadata_ordered.df[metadata_ordered.df[[variable_to_plot]] == member,]) > 2){ # if too few samples, skip plotting ellipse
        ordispider(pca_site_scores,
                   groups = metadata_ordered.df[[variable_to_plot]],
                   # col = variable_colours[member][[1]],
                   col = alpha(variable_colours[member][[1]],point_alpha),
                   show.groups = member,
                   #alpha = .05,
                   label = label_spider)
      }
    }
  }
  if (hasArg(plot_spiders)){
    if (plot_spiders == T){
      plot_spiders_func(F)    
    }
  }
  
  # points(pca_specie_scores, 
  #        cex = 0.8,
  #        col = "red",
  #        bg = "red",
  #        pch = 3,
  # )
  
  if (label_sites == T){
    text(x = pca_site_scores[,1],
         y = pca_site_scores[,2],
         labels = rownames(pca_site_scores),
         cex = .5,
         pos = 2)
  }
  if (label_species == T){
    text(x = pca_specie_scores[,1],
         y = pca_specie_scores[,2],
         labels = rownames(pca_specie_scores),
         cex = .5,
         pos = 2)
    # arrows(x = pca_specie_scores[,1],
    #        y = pca_specie_scores[,2])
  }
  
  if (is.null(legend_x) || is.null(legend_y)){
    legend_x <- x_min + legend_x_offset
    legend_y <- y_max + legend_y_offset
  }
  if (is.null(legend_title)){
    legend_title <- variable_to_plot
  }
  
  if (!is.null(plot_title)){
    title(main = plot_title)
  }
  
  if (include_legend){
    legend(
      # title = bold(variable_to_plot),
      title = as.expression(bquote(bold(.(legend_title)))),
      title.col="black",
      # x = x_min-4,
      # y = y_max-6,
      x = legend_x,
      y = legend_y,
      legend= variable_values,
      pch= unique(all_sample_shapes),
      col= legend_point_outline_colours,
      # col= "black",
      pt.bg = unique(all_sample_colours),
      #bg = "white",
      bty = "n",
      ncol = legend_cols,
      cex = 0.6,
      # pt.cex = 0.6,
      pt.lwd = point_line_thickness,
      y.intersp =1,
      x.intersp =1,
    )  
  }
  
  if (!is.null(filename)){
    dev.off()
  }
}

# ------------------------------------------------------------------------------------------------------------------------
# Testing
# temp_metadata <- subset(metadata.df, Sampletype != "negative")
# temp_data <- otu_rare_filtered.m[,rownames(temp_metadata)]
# temp <- capscale(t(otu_rare_filtered.m)~1, data = temp_metadata, distance = "bray")
# 
# generate_pca(temp, mymetadata = temp_metadata,
#              plot_height = 5, plot_width =5,
#              legend_x = -6, legend_y = 4,
#              point_size = .7, point_line_thickness = .3,point_alpha =.7,
#              legend_title = "Sample type",
#              plot_title = "Both cohorts, all lesion types",
#              limits = c(-5,5,-5,5),
#              plot_spiders = F,
#              plot_ellipses = F,
#              use_shapes = T,
#              ellipse_border_width = .5,
#              label_ellipse = F, ellipse_label_size = .5,
#              colour_palette = my_colour_palette_206_distinct,
#              variable_to_plot = "Sampletype_pooled", legend_cols = 1,
#              variable_colours_available = T,
#              filename = paste0("Result_figures/pcoa_dbrda_plots/both_cohorts_Sampletype_pooled_bray.pdf"))

# metadata_sampletype.df <- subset(metadata.df, Sampletype_pooled == "AK")
# metadata_sampletype.df <- metadata_sampletype.df[order(rownames(metadata_sampletype.df)),]
# otu_rare_sampletype.m <- otu_rare_filtered.m[,colnames(otu_rare_filtered.m) %in% rownames(metadata_sampletype.df)]
# m.pca_sampletype <- capscale(t(otu_rare_filtered.m)~1, data = metadata_sampletype.df, distance = "bray")
# 
# generate_pca(m.pca_sampletype, mymetadata = metadata_sampletype.df,
#              plot_height = 5, plot_width =5,
#              legend_x = -7, legend_y = 6,
#              point_size = .7, point_line_thickness = .3,point_alpha =.7,
#              legend_title = "Cohort",
#              include_legend = T,
#              plot_title = paste0("Sampletype : AK"),
#              # limits = c(-5,5,-5,5),
#              plot_hulls = F,
#              plot_spiders = F,
#              plot_ellipses = F,
#              use_shapes = T,
#              ellipse_border_width = .5,
#              label_ellipse = F, ellipse_label_size = .5,
#              colour_palette = my_colour_palette_206_distinct,
#              variable_to_plot = "Project", legend_cols = 1,
#              variable_colours_available = T,
#              filename = paste0("Result_figures/pcoa_dbrda_plots/AK_cohort_bray.pdf"))

# ------------------------------------------------------------------------------------------------------------------------

# Both cohorts, all sample types 
temp_metadata <- metadata.df #subset(metadata.df, Sampletype != "negative")
temp_data <- otu_rare_clr_filtered.m[,rownames(temp_metadata)]
temp <- rda(t(temp_data), data = temp_metadata) # ~1 makes it unconstrained

generate_pca(temp, mymetadata = temp_metadata,
             plot_height = 5, plot_width =5,
             legend_x = -6, legend_y = 4,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Sample type",
             plot_title = "Both cohorts, all lesion types",
             limits = c(-5,5,-5,5),
             plot_spiders = F,
             plot_ellipses = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Sampletype_pooled", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/both_cohorts_Sampletype_pooled.pdf"))

generate_pca(temp, mymetadata = temp_metadata,
             plot_height = 5, plot_width =5,
             legend_x = -6, legend_y = 4,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Patient",
             plot_title = "Both cohorts, all lesion types",
             limits = c(-5,5,-5,5),
             include_legend = F,
             plot_spiders = F,
             plot_ellipses = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Patient", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/both_cohorts_Patient.pdf"))


generate_pca(temp, mymetadata = temp_metadata,
             plot_height = 5, plot_width =5,
             legend_x = -6, legend_y = 4,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Cohort",
             plot_title = "Both cohorts, all lesion types",
             limits = c(-5,5,-5,5),
             include_legend = T,
             plot_spiders = F,
             plot_ellipses = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Project", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/both_cohorts_Project.pdf"))


# ---------------------------------------------------------------------------------------------------------
# Immunocompetent, all sample types
metadata_immunocompetent.df <- subset(metadata.df, Project == "immunocompetent" & Sampletype != "negative")
metadata_immunocompetent.df <- metadata_immunocompetent.df[order(rownames(metadata_immunocompetent.df)),]
otu_rare_clr_filtered_competent.m <- otu_rare_clr_filtered.m[,colnames(otu_rare_clr_filtered.m) %in% rownames(metadata_immunocompetent.df)]
otu_rare_clr_filtered_competent.m <- otu_rare_clr_filtered_competent.m[,rownames(metadata_immunocompetent.df)]

colnames(otu_rare_clr_filtered_competent.m)[!rownames(metadata_immunocompetent.df) == colnames(otu_rare_clr_filtered_competent.m)]

m.pca_competent <- rda(t(otu_rare_clr_filtered_competent.m), data = metadata_immunocompetent.df)
generate_pca(m.pca_competent, mymetadata = metadata_immunocompetent.df,
             plot_height = 5, plot_width =5,
             legend_x = -6, legend_y = 4,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Lesion type",
             plot_title = "Immunocompetent, all lesion types",
             limits = c(-5,5,-5,5),
             plot_spiders = F,
             plot_ellipses = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Sampletype_pooled", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/immunocompetent_Sampletype_pooled.pdf"))

generate_pca(m.pca_competent, mymetadata = metadata_immunocompetent.df,
             plot_height = 5, plot_width =5,
             legend_x = -6, legend_y = 4,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Patient",
             plot_title = "Immunocompetent, all lesion types",
             limits = c(-5,5,-5,5),
             plot_spiders = F,
             plot_ellipses = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_15,
             variable_to_plot = "Patient", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/immunocompetent_Patient.pdf"))

# ---------------------------------------------------------------------------------------------------------
# Immunocompromised, all sample types
metadata_immunocompromised.df <- subset(metadata.df, Project == "immunocompromised" & Sampletype != "negative")
metadata_immunocompromised.df <- metadata_immunocompromised.df[order(rownames(metadata_immunocompromised.df)),]
otu_rare_clr_filtered_compromised.m <- otu_rare_clr_filtered.m[,colnames(otu_rare_clr_filtered.m) %in% rownames(metadata_immunocompromised.df)]
otu_rare_clr_filtered_compromised.m <- otu_rare_clr_filtered_compromised.m[,rownames(metadata_immunocompromised.df)]

m.pca_compromised <- rda(t(otu_rare_clr_filtered_compromised.m), data = metadata_immunocompromised.df)

# Sampletype_pooled
generate_pca(m.pca_compromised, mymetadata = metadata_immunocompromised.df,
             plot_height = 5, plot_width =5,
             legend_x = -8, legend_y = 7,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Lesion type",
             plot_title = "Immunocompromised, all lesion types",
             limits = c(-7,7,-7,7),
             plot_hulls = F,
             plot_spiders = F,
             plot_ellipses = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Sampletype_pooled", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/immunocompromised_Sampletype_pooled.pdf"))

# Patient
generate_pca(m.pca_compromised, mymetadata = metadata_immunocompromised.df,
             plot_height = 5, plot_width =5,
             legend_x = -6, legend_y = 4,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Patient",
             include_legend = F,
             plot_title = "Immunocompromised, all lesion types",
             limits = c(-7,7,-7,7),
             plot_hulls = F,
             plot_spiders = F,
             plot_ellipses = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Patient", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/immunocompromised_Patient.pdf"))

# Patient_group
generate_pca(m.pca_compromised, mymetadata = metadata_immunocompromised.df,
             plot_height = 5, plot_width =5,
             legend_x = -8, legend_y = 6,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Patient Group",
             plot_title = "Immunocompromised, all lesion types",
             limits = c(-7,7,-7,7),
             include_legend = T,
             plot_spiders = F,
             plot_ellipses = F,
             plot_hulls = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Patient_group", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/immunocompromised_Patient_group.pdf"))

# Gender
generate_pca(m.pca_compromised, mymetadata = metadata_immunocompromised.df,
             plot_height = 5, plot_width =5,
             legend_x = -8, legend_y = 6,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Gender",
             plot_title = "Immunocompromised, all lesion types",
             limits = c(-7,7,-7,7),
             include_legend = T,
             plot_spiders = F,
             plot_ellipses = F,
             plot_hulls = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Gender", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/immunocompromised_Gender.pdf"))

# Number_of_meds
generate_pca(m.pca_compromised, mymetadata = metadata_immunocompromised.df,
             plot_height = 5, plot_width =5,
             legend_x = -8, legend_y = 7,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Number of medications",
             plot_title = "Immunocompromised, all lesion types",
             limits = c(-7,7,-7,7),
             include_legend = T,
             plot_spiders = F,
             plot_ellipses = F,
             plot_hulls = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Number_of_meds", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/immunocompromised_number_of_meds.pdf"))


# Fitzpatrick_skin_type
generate_pca(m.pca_compromised, mymetadata = metadata_immunocompromised.df,
             plot_height = 5, plot_width =5,
             legend_x = -8, legend_y = 7,
             point_size = .7, point_line_thickness = .3,point_alpha =.7,
             legend_title = "Fitzpatrick skin type",
             plot_title = "Immunocompromised, all lesion types",
             limits = c(-7,7,-7,7),
             include_legend = T,
             plot_spiders = F,
             plot_ellipses = F,
             plot_hulls = F,
             use_shapes = T,
             ellipse_border_width = .5,
             label_ellipse = F, ellipse_label_size = .5,
             colour_palette = my_colour_palette_206_distinct,
             variable_to_plot = "Fitzpatrick_skin_type", legend_cols = 1,
             variable_colours_available = T,
             filename = paste0("Result_figures/pcoa_dbrda_plots/immunocompromised_fitzpatrick_skin_type.pdf"))


# ---------------------------------------------------------------------------------------------------------


# Each lesion type, color by cohort and patient
for (sample_type in unique(metadata.df$Sampletype_pooled)){
  metadata_sampletype.df <- subset(metadata.df, Sampletype_pooled == sample_type)
  metadata_sampletype.df <- metadata_sampletype.df[order(rownames(metadata_sampletype.df)),]
  otu_rare_clr_sampletype.m <- otu_rare_clr_filtered.m[,colnames(otu_rare_clr_filtered.m) %in% rownames(metadata_sampletype.df)]
  m.pca_sampletype <- rda(t(otu_rare_clr_sampletype.m), data = metadata_sampletype.df)
  generate_pca(m.pca_sampletype, mymetadata = metadata_sampletype.df,
               plot_height = 5, plot_width =5,
               legend_x = -7, legend_y = 6,
               point_size = .7, point_line_thickness = .3,point_alpha =.7,
               legend_title = "Cohort",
               include_legend = T,
               plot_title = paste0("Sampletype : ", sample_type),
               limits = c(-7,7,-7,7),
               plot_hulls = F,
               plot_spiders = F,
               plot_ellipses = F,
               use_shapes = T,
               ellipse_border_width = .5,
               label_ellipse = F, ellipse_label_size = .5,
               colour_palette = my_colour_palette_206_distinct,
               variable_to_plot = "Project", legend_cols = 1,
               variable_colours_available = T,
               filename = paste0("Result_figures/pcoa_dbrda_plots/",sample_type,"_cohort.pdf"))
  
  
  generate_pca(m.pca_sampletype, mymetadata = metadata_sampletype.df,
               plot_height = 5, plot_width =5,
               legend_x = -7, legend_y = 6,
               point_size = .5, point_line_thickness = .3,point_alpha =.7,
               legend_title = "Patient",
               include_legend = F,
               plot_title = paste0("Sampletype : ", sample_type),
               limits = c(-7,7,-7,7),
               plot_hulls = F,
               plot_spiders = F,
               plot_ellipses = F,
               use_shapes = T,
               ellipse_border_width = .5,
               label_ellipse = F, ellipse_label_size = .5,
               colour_palette = my_colour_palette_206_distinct,
               variable_to_plot = "Patient", legend_cols = 1,
               variable_colours_available = T,
               filename = paste0("Result_figures/pcoa_dbrda_plots/", sample_type, "_Patient.pdf"))
}

# Each lesion type within cohort, color by patient
for (cohort in unique(metadata.df$Project)){
  for (sample_type in unique(metadata.df$Sampletype_pooled)){
    metadata_sampletype.df <- subset(metadata.df, Project == cohort & Sampletype_pooled == sample_type)
    metadata_sampletype.df <- metadata_sampletype.df[order(rownames(metadata_sampletype.df)),]
    otu_rare_clr_sampletype.m <- otu_rare_clr_filtered.m[,colnames(otu_rare_clr_filtered.m) %in% rownames(metadata_sampletype.df)]
    m.pca_sampletype <- rda(t(otu_rare_clr_sampletype.m), data = metadata_sampletype.df)
    generate_pca(m.pca_sampletype, mymetadata = metadata_sampletype.df,
                 plot_height = 5, plot_width =5,
                 legend_x = -10, legend_y = 9,
                 point_size = .7, point_line_thickness = .3, point_alpha =.7,
                 legend_title = "Patient",
                 include_legend = T,
                 plot_title = paste0(cohort, "\nSampletype : ", sample_type, "; Patient"),
                 limits = c(-10,10,-10,10),
                 plot_hulls = F,
                 plot_spiders = F,
                 plot_ellipses = F,
                 use_shapes = T,
                 ellipse_border_width = .5,
                 label_ellipse = F, ellipse_label_size = .5,
                 colour_palette = my_colour_palette_206_distinct,
                 variable_to_plot = "Patient", legend_cols = 1,
                 variable_colours_available = T,
                 filename = paste0("Result_figures/pcoa_dbrda_plots/",cohort, "_",sample_type,"_patient.pdf"))
    
  }
}

# For each patient; all lesion types
# for (patient in unique(metadata.df$Patient)){
#   metadata_patient.df <- subset(metadata.df, Patient == patient)
#   metadata_patient.df <- metadata_patient.df[order(rownames(metadata_patient.df)),]
#   otu_rare_clr_patient.m <- otu_rare_clr_filtered.m[,colnames(otu_rare_clr_filtered.m) %in% rownames(metadata_patient.df)]
#   if (dim(metadata_patient.df)[1] <3) {next} 
#   m.pca_patient <- rda(t(otu_rare_clr_patient.m), data = metadata_patient.df)
#   generate_pca(m.pca_patient, mymetadata = metadata_patient.df,
#                plot_height = 5, plot_width =5,
#                legend_x = -15, legend_y = 10,
#                point_size = .7, point_line_thickness = .3,point_alpha =.7,
#                legend_title = "Lesion type",
#                include_legend = T,
#                plot_title = paste0("Patient : ", patient),
#                limits = c(-15,15,-15,15),
#                plot_hulls = F,
#                plot_spiders = F,
#                plot_ellipses = F,
#                use_shapes = T,
#                ellipse_border_width = .5,
#                label_ellipse = F, ellipse_label_size = .5,
#                colour_palette = my_colour_palette_206_distinct,
#                variable_to_plot = "Sampletype_pooled", legend_cols = 1,
#                variable_colours_available = T,
#                filename = paste0("Result_figures/pcoa_dbrda_plots/",patient,".pdf"))
# }

# ------------------------------------------------------------------------
# ------------------------------------------------------------------------
# ------------------------------------------------------------------------
# PERMANOVA analysis

# Permutational Multivariate Analysis of Variance (PERMANOVA) can be used to 
# determine if the structure of the microbial communities is significantly different between
# environmental variables. This is done using the adonis function from Vegan with a distance metric, e.g. Bray-Curtis, and a
# specified number of permutations, e.g. 10,000.
# The analysis measures the degree each environmental variable affects the community composition and indicates 
# the significance of that effect on beta diversity (described by p-values and R2 values). 
# The R2 value corresponds to the proportion of variability observed in the dissimilarity.


# The function below will only calculate the the significance of individual variables
run_permanova <- function(my_community_data, my_metadata, my_variables){
  stat_sig_table <- NULL
  
  # Remove NA entries from the metadata
  for (var_name in my_variables) {
    result <- adonis(my_community_data~get(var_name),data = my_metadata, permu=999,method="euclidean")
    SumOfSqs <- round(result$aov.tab$SumsOfSqs[1], 3)
    meanSqs <- round(result$aov.tab$MeanSqs[1], 3)
    F.model <- round(result$aov.tab$F.Model[1], 3)
    R2 <- round(result$aov.tab$R2[1], 3)
    p_value <- round(result$aov.tab$`Pr(>F)`[1], 5)
    stat_sig_table <- rbind(stat_sig_table, data.frame(var_name,
                                                       SumOfSqs,
                                                       meanSqs,
                                                       F.model,
                                                       R2,
                                                       p_value))
  }
  names(stat_sig_table) <- c("Variable", "SumOfSqs","MeanSqs","F.Model","R2","P-value")
  stat_sig_table <- stat_sig_table[order(stat_sig_table$"P-value"),]
  stat_sig_table
}

run_permanova_custom <- function(my_metadata, my_formula){
  stat_sig_table <- NULL
  result <- adonis(my_formula,data = my_metadata, permu=999,method="euclidean")
  for (r in rownames(result$aov.tab)){
    variable <- r
    Degress_of_freedom <- result$aov.tab[r,]$Df[1]
    SumOfSqs <- round(result$aov.tab[r,]$SumsOfSqs[1], 3)
    meanSqs <- round(result$aov.tab[r,]$MeanSqs[1], 3)
    F.model <- round(result$aov.tab[r,]$F.Model[1], 3)
    R2 <- round(result$aov.tab[r,]$R2[1], 3)
    p_value <- round(result$aov.tab[r,]$`Pr(>F)`[1], 5)
    stat_sig_table <- rbind(stat_sig_table, data.frame(variable,
                                                       Degress_of_freedom,
                                                       SumOfSqs,
                                                       meanSqs,
                                                       F.model,
                                                       R2,
                                                       p_value))
  }
  print(result)
  names(stat_sig_table) <- c("Term","Df", "SumOfSqs","MeanSqs","F.Model","R2","Pr(>F)")
  stat_sig_table <- stat_sig_table[order(stat_sig_table$"Pr(>F)"),]
  stat_sig_table
}

# Both cohorts
# adonis(t(otu_rare_clr_filtered.m)~Patient+Project +Sampletype_pooled+Patient:Sampletype_pooled,data = metadata.df, permu=999,method="euclidean")

# Immunocompromised
metadata_immunocompromised.df <- subset(metadata.df, Project == "immunocompromised" & Sampletype != "negative")
metadata_immunocompromised.df <- metadata_immunocompromised.df[order(rownames(metadata_immunocompromised.df)),]
otu_rare_clr_filtered_compromised.m <- otu_rare_clr_filtered.m[,colnames(otu_rare_clr_filtered.m) %in% rownames(metadata_immunocompromised.df)]
otu_rare_clr_filtered_compromised.m <- otu_rare_clr_filtered_compromised.m[,rownames(metadata_immunocompromised.df)]

# result <- adonis(t(otu_rare_clr_filtered_compromised.m)~Patient+Sampletype_pooled+Patient:Sampletype_pooled,data = metadata_immunocompromised.df, permu=999,method="euclidean")
permanova_results_immunocompromised <- run_permanova_custom(my_metadata = metadata_immunocompromised.df,
                     my_formula = as.formula(t(otu_rare_clr_filtered_compromised.m)~Patient+Sampletype_pooled+Patient:Sampletype_pooled))

permanova_results_immunocompromised <- run_permanova_custom(my_metadata = metadata_immunocompromised.df,
                                                            my_formula = as.formula("t(otu_rare_clr_filtered_compromised.m)~Patient+Sampletype_pooled+Gender+Patient_group + Number_of_meds"))

# Immunocompetent, all sample types
metadata_immunocompetent.df <- subset(metadata.df, Project == "immunocompetent" & Sampletype != "negative")
metadata_immunocompetent.df <- metadata_immunocompetent.df[order(rownames(metadata_immunocompetent.df)),]
otu_rare_clr_filtered_competent.m <- otu_rare_clr_filtered.m[,colnames(otu_rare_clr_filtered.m) %in% rownames(metadata_immunocompetent.df)]
otu_rare_clr_filtered_competent.m <- otu_rare_clr_filtered_competent.m[,rownames(metadata_immunocompetent.df)]

permanova_results_immunocompetent <- run_permanova_custom(my_metadata = metadata_immunocompetent.df,
                                                            my_formula = as.formula(t(otu_rare_clr_filtered_competent.m)~Patient+Sampletype_pooled+Patient:Sampletype_pooled))

# Both cohorts, all samples types
permanova_results_both_cohorts <- run_permanova_custom(my_metadata = metadata.df,
                                                       my_formula = as.formula(t(otu_rare_clr_filtered.m)~Patient+Sampletype_pooled+Patient:Sampletype_pooled))

write.csv(permanova_results_both_cohorts,file="Result_tables/stats_various/PERMANOVA_both_cohorts.csv",row.names = F)
write.csv(permanova_results_immunocompromised,file="Result_tables/stats_various/PERMANOVA_immunocompromised.csv",row.names = F)
write.csv(permanova_results_immunocompetent,file="Result_tables/stats_various/PERMANOVA_immunocompetent.csv",row.names = F)

